library/core/tests/array.rs

```diff
@@ -301,3 +301,31 @@ fn array_map() {
     let b = a.map(|v| v as u64);
     assert_eq!(b, [1, 2, 3]);
 }
+
+#[test]
+fn array_map_drop_safety() {
+    static mut DROPPED: usize = 0;
```

Comment:
I'd prefer using a static `AtomicUsize` or even an `Rc<Cell<usize>>` in each `DropCounter` here to avoid `unsafe` blocks below. The performance here really doesn't matter and less `unsafe` is better.